src/Compilers/CSharp/Test/Emit2/Semantics/PatternMatchingTests_ListPatterns.cs

```diff
@@ -8936,4 +8938,554 @@ public void NotExhaustive_LongList()
+              IL_0017:  ldc.i4.s   41
+              IL_0019:  bne.un.s   IL_0028
+              IL_001b:  ldarg.1
+              IL_001c:  ldc.i4.0
+              IL_001d:  ldloc.1
+              IL_001e:  ldc.i4.1
+              IL_001f:  sub
+              IL_0020:  callvirt   "string string.Substring(int, int)"
+              IL_0025:  stloc.0
+              IL_0026:  ldloc.0
+              IL_0027:  ret
+              IL_0028:  ldnull
+              IL_0029:  ret
+            }
+            """);
+    }
+
+    [Theory, WorkItem("https://github.com/dotnet/roslyn/issues/60091")]
+    [InlineData("var")]
+    [InlineData("ReadOnlySpan<string>")]
+    public void SlicePattern_FirstDeclarationPatternOfSameType_ReadOnlySpan_CodeGen(string typeSpecifier)
+    {
+        var source = $$"""
+            using System;
+
+            class C
+            {
+                ReadOnlySpan<string> M(ReadOnlySpan<string> s)
+                {
+                    if (s is [..{{typeSpecifier}} start, ")"])
+                    {
+                        return start;
+                    }
+
+                    return default;
+                }
+            }
+            """;
+
+        var comp = CreateCompilationWithIndexAndRangeAndSpan(new[] { source, TestSources.GetSubArray });
+        comp.VerifyEmitDiagnostics();
+
+        CompileAndVerify(comp, verify: Verification.FailsILVerify).VerifyIL("C.M", """
+            {
+              // Code size       59 (0x3b)
+              .maxstack  4
+              .locals init (System.ReadOnlySpan<string> V_0, //start
+                            int V_1,
+                            System.ReadOnlySpan<string> V_2)
+              IL_0000:  ldarga.s   V_1
+              IL_0002:  call       "int System.ReadOnlySpan<string>.Length.get"
+              IL_0007:  stloc.1
+              IL_0008:  ldloc.1
+              IL_0009:  ldc.i4.1
+              IL_000a:  blt.s      IL_0031
+              IL_000c:  ldarga.s   V_1
+              IL_000e:  ldloc.1
+              IL_000f:  ldc.i4.1
+              IL_0010:  sub
+              IL_0011:  call       "ref readonly string System.ReadOnlySpan<string>.this[int].get"
+              IL_0016:  ldind.ref
+              IL_0017:  ldstr      ")"
+              IL_001c:  call       "bool string.op_Equality(string, string)"
+              IL_0021:  brfalse.s  IL_0031
+              IL_0023:  ldarga.s   V_1
+              IL_0025:  ldc.i4.0
+              IL_0026:  ldloc.1
+              IL_0027:  ldc.i4.1
+              IL_0028:  sub
+              IL_0029:  call       "System.ReadOnlySpan<string> System.ReadOnlySpan<string>.Slice(int, int)"
+              IL_002e:  stloc.0
+              IL_002f:  ldloc.0
+              IL_0030:  ret
+              IL_0031:  ldloca.s   V_2
+              IL_0033:  initobj    "System.ReadOnlySpan<string>"
+              IL_0039:  ldloc.2
+              IL_003a:  ret
+            }
+            """);
+    }
```

Comment:
All the added tests use pre-existing types (arrays, string, ROS) so the change is only observable through IL inspection. 
Please add some tests with custom collection types where the different properties/methods are instrumented to print something, then we can directly verify effect on order of evaluation.
See `SlicePattern_Subpattern` for an example (`Length` property logs itself).

Reply:
The first one is done.

> I don't think that scenario will get past initial binding, so shouldn't reach the modified code, but it may be good to check

Correct, that's a binding problem. And I am 99.99% sure these cases are already tested in any meaningful way possible somewhere in the binding tests. I think adding them in this PR will just create unnecessary duplicates and bloat the size of code to review. So left this one untouched. If you really feal strong, that I am wrong and such tests are actually needed here, LMK

Reply:
@jcouv I need your final word here